File: Circles.Profiles.Sdk/NamespaceWriter.cs
using System.Text;
using System.Text.Json;
using Circles.Profiles.Interfaces;
using Circles.Profiles.Models;
using Circles.Profiles.Models.Core;
using Circles.Profiles.Sdk.Utils;
using Nethereum.Hex.HexConvertors.Extensions;

namespace Circles.Profiles.Sdk;

/// <summary>
/// Writes to one <em>(ownerAvatar, namespaceKey)</em> log.
/// If the same <paramref name="logicalName"/> is written again
/// the newer link <b>replaces</b> the older entry <i>inside the head chunk</i>.
/// Across chunk rotations the profile‑level index ensures random‑access always
/// resolves to the most recent occurrence while older links stay in history.
/// </summary>
public sealed class NamespaceWriter : INamespaceWriter
{
    private readonly Profile _ownerProfile;
    private readonly string _nsKeyLower; // recipient‑addr lower‑cased
    private readonly IIpfsStore _ipfs;
    private readonly ILinkSigner _signer;

    private NameIndexDoc _index = new(); // hydrated in CreateAsync
    private NamespaceChunk _head = new(); // idem

    private NamespaceWriter(Profile ownerProfile,
        string namespaceKey,
        IIpfsStore ipfs,
        ILinkSigner signer)
    {
        _ownerProfile = ownerProfile;
        _nsKeyLower = namespaceKey.ToLowerInvariant();
        _ipfs = ipfs;
        _signer = signer;
    }

    /// <summary>
    /// Asynchronously loads the existing index/chunk state and returns a ready‑to‑use writer.
    /// </summary>
    public static async Task<NamespaceWriter> CreateAsync(
        Profile ownerProfile,
        string namespaceKey,
        IIpfsStore ipfs,
        ILinkSigner signer,
        CancellationToken ct = default)
    {
        var w = new NamespaceWriter(ownerProfile, namespaceKey, ipfs, signer);

        bool hasIndex = ownerProfile.Namespaces.TryGetValue(w._nsKeyLower, out var idxCid);
        if (hasIndex)
        {
            w._index = await Helpers.LoadIndex(idxCid, ipfs, ct);
            w._head = await Helpers.LoadChunk(w._index.Head, ipfs, ct);
        }

        return w;
    }

    /* ───────────── single helpers ────────────────────────────────────── */

    public async Task<CustomDataLink> AddJsonAsync(string name,
        string json,
        string pk,
        CancellationToken ct = default)
    {
        string cid = await _ipfs.AddStringAsync(json, pin: true, ct);
        return await AttachExistingCidAsync(name, cid, pk, ct);
    }

    public async Task<CustomDataLink> AttachExistingCidAsync(
        string name,
        string cid,
        string pk,
        CancellationToken ct = default)
    {
        bool nameEmpty = string.IsNullOrWhiteSpace(name);
        if (nameEmpty)
        {
            throw new ArgumentNullException(nameof(name));
        }

        bool cidEmpty = string.IsNullOrWhiteSpace(cid);
        if (cidEmpty)
        {
            throw new ArgumentNullException(nameof(cid));
        }

        bool pkEmpty = string.IsNullOrWhiteSpace(pk);
        if (pkEmpty)
        {
            throw new ArgumentNullException(nameof(pk));
        }

        var draft = new CustomDataLink
        {
            Name = name,
            Cid = cid,
            ChainId = Helpers.DefaultChainId,
            SignedAt = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
            Nonce = CustomDataLink.NewNonce(),
            Encrypted = false
        };

        CustomDataLink signed = _signer.Sign(draft, pk);

        await PersistAsync([signed], ct);
        return signed;
    }

    /* bulk -------------------------------------------------------- */

    public async Task<IReadOnlyList<CustomDataLink>> AddJsonBatchAsync(
        IEnumerable<(string name, string json)> items,
        string pk,
        CancellationToken ct = default)
    {
        bool pkEmpty = string.IsNullOrWhiteSpace(pk);
        if (pkEmpty)
        {
            throw new ArgumentNullException(nameof(pk));
        }

        var vec = new List<CustomDataLink>();

        var itemsArray = items.ToArray();

        bool anyNameMissing = itemsArray.Any(o => string.IsNullOrWhiteSpace(o.name));
        if (anyNameMissing)
        {
            throw new ArgumentException("At least one of the items in the list doesn't have a name.", nameof(items));
        }

        foreach (var (n, j) in itemsArray)
        {
            ct.ThrowIfCancellationRequested();
            string cid = await _ipfs.AddStringAsync(j, pin: true, ct);

            vec.Add(new CustomDataLink
            {
                Name = n,
                Cid = cid,
                ChainId = Helpers.DefaultChainId,
                SignedAt = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
                Nonce = CustomDataLink.NewNonce(),
                Encrypted = false
            });
        }

        vec = vec.Select(l => _signer.Sign(l, pk)).ToList();
        await PersistAsync(vec, ct);
        return vec;
    }

    public async Task<IReadOnlyList<CustomDataLink>> AttachCidBatchAsync(
        IEnumerable<(string name, string cid)> items,
        string pk,
        CancellationToken ct = default)
    {
        bool pkEmpty = string.IsNullOrWhiteSpace(pk);
        if (pkEmpty)
        {
            throw new ArgumentNullException(nameof(pk));
        }

        var vec = items.Select(i =>
        {
            bool nameEmpty = string.IsNullOrWhiteSpace(i.name);
            if (nameEmpty)
            {
                throw new ArgumentNullException(nameof(i.name));
            }

            bool cidEmpty = string.IsNullOrWhiteSpace(i.cid);
            if (cidEmpty)
            {
                throw new ArgumentNullException(nameof(i.cid));
            }

            return new CustomDataLink
            {
                Name = i.name,
                Cid = i.cid,
                SignedAt = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
                Nonce = CustomDataLink.NewNonce(),
                Encrypted = false
            };
        }).ToList();

        vec = vec.Select(l => _signer.Sign(l, pk)).ToList();
        await PersistAsync(vec, ct);
        return vec;
    }

    /* internals --------------------------------------------------- */

    private async Task PersistAsync(IEnumerable<CustomDataLink> newLinks, CancellationToken ct)
    {
        foreach (var link in newLinks)
        {
            ct.ThrowIfCancellationRequested();

            bool full = _head.Links.Count >= Helpers.ChunkMaxLinks;
            if (full)
            {
                string closedCid = await Helpers.SaveChunk(_head, _ipfs, ct);

                foreach (var l in _head.Links)
                {
                    _index.Entries[l.Name] = closedCid;
                }

                _head = new NamespaceChunk { Prev = closedCid };
            }

            int idx = _head.Links.FindIndex(l => l.Name.Equals(link.Name, StringComparison.OrdinalIgnoreCase));
            bool exists = idx >= 0;
            if (exists)
            {
                _head.Links[idx] = link;
            }
            else
            {
                _head.Links.Add(link);
            }
        }

        string headCid = await Helpers.SaveChunk(_head, _ipfs, ct);

        foreach (var l in _head.Links)
        {
            _index.Entries[l.Name] = headCid;
        }

        _index.Head = headCid;

        string indexJson = JsonSerializer.Serialize(_index, Models.JsonSerializerOptions.JsonLd);
        string indexCid = await _ipfs.CalcCidAsync(
            Encoding.UTF8.GetBytes(indexJson), ct);

        _ownerProfile.Namespaces[_nsKeyLower] = indexCid;

        await _ipfs.AddStringAsync(indexJson, pin: true, ct);
    }

    /// <summary>
    /// Inserts a <b>pre‑signed</b> link after validating
    ///   • its signature
    ///   • signerAddress == namespace key (this writer’s <c>_nsKeyLower</c>)
    /// </summary>
    public async Task<CustomDataLink> AcceptSignedLinkAsync(
        CustomDataLink signedLink,
        Profile operatorProfile,
        IChainApi chain,
        CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(signedLink);
        ArgumentNullException.ThrowIfNull(operatorProfile);

        bool sameNamespace = signedLink.SignerAddress.Equals(_nsKeyLower, StringComparison.OrdinalIgnoreCase);
        if (!sameNamespace)
        {
            throw new InvalidOperationException(
                $"link signer ({signedLink.SignerAddress}) must equal namespace key ({_nsKeyLower})");
        }

        var verifier = new DefaultSignatureVerifier(chain);

        byte[] payload = CanonicalJson.CanonicaliseWithoutSignature(signedLink);
        byte[] hash = Sha3.Keccak256Bytes(payload);
        byte[] sig = signedLink.Signature.HexToByteArray();

        bool primaryOk = await verifier.VerifyAsync(
            hash,
            signedLink.SignerAddress,
            sig,
            ct).ConfigureAwait(false);

        bool needsBytesFallback = !primaryOk;
        if (needsBytesFallback)
        {
            bool bytesOk = await verifier.Verify1271WithBytesAsync(
                payload,
                signedLink.SignerAddress,
                sig,
                ct).ConfigureAwait(false);

            if (!bytesOk)
            {
                throw new InvalidOperationException("invalid signature on supplied link");
            }
        }

        string code = await chain.GetCodeAsync(signedLink.SignerAddress, ct).ConfigureAwait(false);
        bool isEoa = string.Equals(code, "0x", StringComparison.OrdinalIgnoreCase);
        if (isEoa)
        {
            var sigObj = Nethereum.Signer.EthECDSASignatureFactory.ExtractECDSASignature(signedLink.Signature);
            var rec = Nethereum.Signer.EthECKey.RecoverFromSignature(sigObj, hash)
                      ?? throw new InvalidOperationException("failed to recover public key");

            string fp = SigningKeyUtils.ComputeFingerprint(rec);
            bool keyValid = SigningKeyUtils.IsFingerprintValid(fp, operatorProfile, signedLink.SignedAt);
            if (!keyValid)
            {
                throw new InvalidOperationException("signing key is expired / revoked");
            }
        }

        await PersistAsync([signedLink], ct).ConfigureAwait(false);
        return signedLink;
    }
}

File: js/src/NamespaceWriter.ts
import { CustomDataLink } from './interfaces/CustomDataLink';
import { IIpfsStore } from './interfaces/IIpfsStore';
import { ILinkSigner } from './interfaces/ILinkSigner';
import { INamespaceWriter } from './interfaces/INamespaceWriter';
import { NameIndexDoc } from './interfaces/NameIndexDoc';
import { NamespaceChunk } from './interfaces/NamespaceChunk';
import { Profile } from './interfaces/Profile';
import { Helpers } from './utils/Helpers';

/**
 * Writes to one (ownerAvatar, namespaceKey) log.
 * If the same logicalName is written again, the newer link replaces
 * the older entry inside the head chunk.
 */
export class NamespaceWriter implements INamespaceWriter {
  private readonly ownerProfile: Profile;
  private readonly nsKeyLower: string;
  private readonly ipfs: IIpfsStore;
  private readonly signer: ILinkSigner;
  
  private index: NameIndexDoc = { head: '', entries: {} };
  private head: NamespaceChunk = { prev: null, links: [] };
  
  private constructor(
    ownerProfile: Profile,
    namespaceKey: string,
    ipfs: IIpfsStore,
    signer: ILinkSigner
  ) {
    this.ownerProfile = ownerProfile;
    this.nsKeyLower = namespaceKey.toLowerCase();
    this.ipfs = ipfs;
    this.signer = signer;
  }
  
  /**
   * Asynchronously loads the existing index/chunk state and returns a ready-to-use writer.
   * 
   * @param ownerProfile The owner's profile
   * @param namespaceKey The namespace key
   * @param ipfs The IPFS store
   * @param signer The link signer
   * @returns A namespace writer
   */
  public static async createAsync(
    ownerProfile: Profile,
    namespaceKey: string,
    ipfs: IIpfsStore,
    signer: ILinkSigner
  ): Promise<NamespaceWriter> {
    const writer = new NamespaceWriter(ownerProfile, namespaceKey, ipfs, signer);
    
    if (ownerProfile.namespaces && ownerProfile.namespaces[writer.nsKeyLower]) {
      const idxCid = ownerProfile.namespaces[writer.nsKeyLower];
      writer.index = await Helpers.loadIndex(idxCid, ipfs);
      writer.head = await Helpers.loadChunk(writer.index.head, ipfs);
    }
    
    return writer;
  }
  
  /**
   * Adds JSON content to the namespace.
   * 
   * @param name Logical name for the content
   * @param json JSON content to add
   * @param pk Private key for signing
   * @returns The created link
   */
  public async addJsonAsync(
    name: string,
    json: string,
    pk: string
  ): Promise<CustomDataLink> {
    const cid = await this.ipfs.addJsonAsync(json, true);
    return this.attachExistingCidAsync(name, cid, pk);
  }
  
  /**
   * Attaches an existing CID to the namespace.
   * 
   * @param name Logical name for the content
   * @param cid IPFS CID to attach
   * @param pk Private key for signing
   * @returns The created link
   */
  public async attachExistingCidAsync(
    name: string,
    cid: string,
    pk: string
  ): Promise<CustomDataLink> {
    if (!name || name.trim() === '') throw new Error('Name is required');
    if (!cid || cid.trim() === '') throw new Error('CID is required');
    if (!pk || pk.trim() === '') throw new Error('Private key is required');
    
    const draft: CustomDataLink = {
      name,
      cid,
      chainId: Helpers.DefaultChainId,
      signedAt: Math.floor(Date.now() / 1000),
      nonce: this.generateNonce(),
      encrypted: false
    };
    
    const signed = this.signer.sign(draft, pk);
    
    await this.persistAsync([signed]);
    return signed;
  }
  
  /**
   * Adds multiple JSON contents to the namespace.
   * 
   * @param items Array of name-json pairs
   * @param pk Private key for signing
   * @returns Array of created links
   */
  public async addJsonBatchAsync(
    items: Array<[string, string]>,
    pk: string
  ): Promise<ReadonlyArray<CustomDataLink>> {
    if (!pk || pk.trim() === '') {
      throw new Error('Private key is required');
    }
    
    const links: CustomDataLink[] = [];
    
    for (const [name, json] of items) {
      if (!name || name.trim() === '') {
        throw new Error('At least one of the items in the list doesn\'t have a name');
      }
      
      const cid = await this.ipfs.addJsonAsync(json, true);
      
      links.push({
        name,
        cid,
        chainId: Helpers.DefaultChainId,
        signedAt: Math.floor(Date.now() / 1000),
        nonce: this.generateNonce(),
        encrypted: false
      });
    }
    
    const signedLinks = links.map(link => this.signer.sign(link, pk));
    await this.persistAsync(signedLinks);
    return signedLinks;
  }
  
  /**
   * Attaches multiple existing CIDs to the namespace.
   * 
   * @param items Array of name-cid pairs
   * @param pk Private key for signing
   * @returns Array of created links
   */
  public async attachCidBatchAsync(
    items: Array<[string, string]>,
    pk: string
  ): Promise<ReadonlyArray<CustomDataLink>> {
    if (!pk || pk.trim() === '') {
      throw new Error('Private key is required');
    }
    
    const links = items.map(([name, cid]) => {
      if (!name || name.trim() === '') {
        throw new Error('Name is required');
      }
      if (!cid || cid.trim() === '') {
        throw new Error('CID is required');
      }
      
      return {
        name,
        cid,
        chainId: Helpers.DefaultChainId,
        signedAt: Math.floor(Date.now() / 1000),
        nonce: this.generateNonce(),
        encrypted: false
      };
    });
    
    const signedLinks = links.map(link => this.signer.sign(link, pk));
    await this.persistAsync(signedLinks);
    return signedLinks;
  }
  
  /**
   * Persists links to IPFS and updates the namespace.
   * 
   * @param newLinks Links to persist
   */
  private async persistAsync(newLinks: CustomDataLink[]): Promise<void> {
    for (const link of newLinks) {
      // Rotate if full
      if (this.head.links.length >= Helpers.ChunkMaxLinks) {
        const closedCid = await Helpers.saveChunk(this.head, this.ipfs);
        
        for (const l of this.head.links) {
          this.index.entries[l.name] = closedCid;
        }
        
        this.head = { prev: closedCid, links: [] };
      }
      
      // Up-sert
      const idx = this.head.links.findIndex(
        l => l.name.toLowerCase() === link.name.toLowerCase()
      );
      
      if (idx >= 0) {
        this.head.links[idx] = link;
      } else {
        this.head.links.push(link);
      }
    }
    
    // Flush
    const headCid = await Helpers.saveChunk(this.head, this.ipfs);
    
    for (const l of this.head.links) {
      this.index.entries[l.name] = headCid;
    }
    
    this.index.head = headCid;
    
    const indexCid = await Helpers.saveIndex(this.index, this.ipfs);
    
    if (!this.ownerProfile.namespaces) {
      this.ownerProfile.namespaces = {};
    }
    
    this.ownerProfile.namespaces[this.nsKeyLower] = indexCid;
  }
  
  /**
   * Generates a new random nonce.
   * 
   * @returns A random nonce string
   */
  private generateNonce(): string {
    const bytes = new Uint8Array(16);
    crypto.getRandomValues(bytes);
    return '0x' + Array.from(bytes)
      .map(b => b.toString(16).padStart(2, '0'))
      .join('');
  }
}
